### PR TITLE
Oracle for labelling

### PR DESCRIPTION
**Pull request by moreymat**
This series of commits aims at providing an oracle for labelling that works as expected.

### CODE REVIEW

**Comment by kowey** [comment]
Thanks for the sanity check! I'll just push this up.

**Comment by kowey** [comment]
Oh! Not pushing. Phew.

File: attelo/decoding/control.py [hunk 1]
```diff
@@ -60,10 +60,14 @@ def _predict_relate(dpack, models):
     and an list of labels
     """
     if models.relate == 'oracle':
-        idxes = dpack.target
+        unrelated = dpack.label_number(UNRELATED)
         # pylint: disable=no-member
-        probs = np.ones(idxes.shape)
+        # treat unrelated as unknown (to avoid cutting links that may
+        # have been inserted by an non-oracle attachment label)
+        always_related = np.vectorize(lambda x: 0 if x == unrelated else x)
```
Comment (moreymat) [inline attelo/decoding/control.py]: I find the names quite unclear or misleading:
* `always_related`: `oracle_labeller` ?
* in the anonymous function, `x`: `lbl` ?
* `idxes`: `lbls`
* `probs`: `scores` (which is a general replacement to do)

File: attelo/decoding/control.py [hunk 2]
```diff
@@ -60,10 +60,14 @@ def _predict_relate(dpack, models):
     and an list of labels
     """
     if models.relate == 'oracle':
-        idxes = dpack.target
+        unrelated = dpack.label_number(UNRELATED)
```
Comment (moreymat) [inline attelo/decoding/control.py]: We should probably add `unk_lbl = 0`, or even better, `unk_lbl = dpack.label_number(UNKNOWN)` with the unknown label automatically added to the label encoder, as it is done in `educe`.
Then the anonymous function below should be adapted: ...`lambda lbl: unk_lbl if lbl == unrelated else x`.

**Comment by moreymat** [comment]
Code review finished.

File: attelo/decoding/control.py (hunk 1 above)
Comment (kowey) [inline attelo/decoding/control.py]: Hmm, OK. `s/x/lbl/` (maybe l if too long), `idxes/labels/` and indeed `s/probs/scores/`.
I kind of would prefer to hang on to `always_related` or in any case, a name that communicates a bit the low-level intent.  Any other suggestions for how I might rename this?

Comment (moreymat) [inline attelo/decoding/control.py]: `map_unrel_to_unk` ?

Comment (kowey) [inline attelo/decoding/control.py]: Oh wait, here they literally are probabilities.  So maybe I should keep probs here, and change to scores elsewhere.  Or maybe that's a recipe for pain.

Comment (moreymat) [inline attelo/decoding/control.py]: It does not seem to me that they literally are probabilities here. We haven't used classifiers that do not output probabilities, e.g. sklearn's perceptron, for labelling yet but it makes perfect sense to try that too.

File: attelo/decoding/control.py (hunk 2 above)
Comment (kowey) [inline attelo/decoding/control.py]: OK! Now updated, hopefully clearer now.
I'll defer the decision on probs/scores etc later, but I've taken the liberty of renaming prob_distrib to candidates.  Will push this in an hour or so.

**Comment by moreymat** [comment]
LGTM

**Comment by kowey** [comment]
Thanks for the review!  OK, I think I can feel better about renaming probs to scores with your comment. Am a bit iffy.  Will likely be sweeping through this later anyway when I work on encyclopaedic decoders (ie. where we just shove everything to the decoders and say “yeah, have fun!”)

**Comment by moreymat** [comment]
I still don't get what encyclopaedic decoders are, but heh.

**Pull request merged**